core/src/main/scala/quasar/destination/avalanche/AvalancheDestination.scala

```diff
@@ -0,0 +1,252 @@
+
+      _ <- debug(s"Table creation query:\n${createQuery.sql}")
+
+      _ <- debugRedacted(s"Load query:\n${loadQuery.sql}")
+
+      _ <- debug(s"Drop table query:\n${dropQuery.sql}")
+
+      // drop down to raw JDBC to execute the load query.
+      // Otherwise Avalanche refuses to load
+      load = raw { cn =>
+        val statement = cn.createStatement()
+
+        statement.execute(loadQuery.sql)
+
+        val count = statement.getUpdateCount
+
+        statement.close
+
+        count
+      }
+
+      count <- (dropQuery.run *> createQuery.run *> load).transact(xa)
+
+      _ <- debug(s"Finished load")
+
+      _ <- debug(s"Load result count: $count")
+
+      _ <- debug(s"Finished table copy")
+
+    } yield ()
+
+  private def upload(bytes: Stream[F, Byte]): F[String] =
+    for {
+      freshNameSuffix <- Sync[F].delay(UUID.randomUUID().toString)
+      freshName = s"reform-$freshNameSuffix"
+
+      _ <- refreshToken
+
+      containerURL <- refContainerURL.get
+
+      put = AzurePutService.mk[F](containerURL.value)
+
+      _ <- debug(s"Starting upload of $freshName")
+
+      _ <- put((BlobPath(List(PathElem(freshName))), bytes))
+
+      _ <- debug(s"Finished upload of $freshName")
+
+    } yield freshName
+
+  private def ensureSingleSegment(r: ResourcePath): F[FileName] =
+    r match {
+      case file /: ResourcePath.Root => FileName(file).pure[F]
+      case _ => MonadResourceErr[F].raiseError(ResourceError.notAResource(r))
+    }
+
+  private def copyQuery(tableName: String, fileName: String): Fragment = {
+    val table = removeQuotes(tableName)
+    val clientId = removeQuotes(config.azureCredentials.clientId.value)
+    val clientSecret = removeQuotes(config.azureCredentials.clientSecret.value)
+
+    fr"COPY" ++ Fragment.const0(s""""$table"""") ++ fr0"() VWLOAD FROM " ++ abfsPath(fileName) ++
+      fr"WITH" ++ pairs(
+      fr"AZURE_CLIENT_ENDPOINT" -> authEndpoint(config.azureCredentials.tenantId),
+      fr"AZURE_CLIENT_ID" -> Fragment.const(s"'$clientId'"),
+      fr"AZURE_CLIENT_SECRET" -> Fragment.const(s"'$clientSecret'"),
+      fr"FDELIM" -> fr"','",
+      fr"QUOTE" -> fr"""'"'""")
+  }
+
+  private def authEndpoint(tid: TenantId): Fragment = {
+    val tenant = removeQuotes(tid.value)
+
+    Fragment.const(s"'https://login.microsoftonline.com/$tenant/oauth2/token'")
+  }
+
+  private def pairs(ps: (Fragment, Fragment)*): Fragment =
+    (ps map {
+      case (lhs, rhs) => lhs ++ fr"=" ++ rhs
+    }).toList.intercalate(fr",")
+
+  private def abfsPath(file: String): Fragment = {
+    val container = removeQuotes(config.containerName.value)
+    val account = removeQuotes(config.accountName.value)
+    val fileName = removeQuotes(file)
+
+    Fragment.const(
+      s"'abfs://$container@$account.dfs.core.windows.net/$fileName'")
+  }
+
+  private def createTableQuery(tableName: String, columns: NonEmptyList[Fragment]): Fragment = {
+    val table = removeQuotes(tableName)
```

Comment:
Can you add a comment/link mentioning what valid table names for this db?
Is this everything that needs to happen to ensure it's a valid table name?

Reply:
I'm not trying to validate the name. I'm trying to avoid simple SQL injection by deleting quotes from identifiers and credentials. Unfortunately Ingres doesn't support prepared statements for DDL or `COPY VWLOAD` statements. They don't provide a quoting implementation either, so this is the best we can do.